### DISCUSSION TITLE
API specifications

### DISCUSSION THREAD

**Discussion by joulev** [joulev/ezkomment#97] [Questions] [answered]
@VietAnh1010 What is the specifications of all API that we currently have?

Something like [this](https://vercel.com/docs/rest-api#endpoints).

**Comment by VietAnh1010** [comment]
@joulev Will write the API specifications tonight.

**Comment by VietAnh1010** [comment, marked as answer]
# API

Ezkomment's API follows REST conventions. All endpoints live under the URL `https://ezkdev.joulev.dev/api`.

# Basics

## Content Type

All requests must be encoded as JSON with the `Content-Type: application/json` header. Responses from the API are encoded as JSON as well. Other encodings are not supported.

## Successful Responses

All successful responses will contain a `message` and optional `data`, which is a JSON object,
or an JSON array in case multiple records are read.

```json
{
  "message": "Successfully get user's information",
  "data": {
    "uid": "some uid",
    "displayName": "some name",
    "email": "some email"
  }
}
```

## Failed Responses

All failed responses will contain a `message` and an `error` that contains the error message. (This
design is planned to be changed).

# Endpoints

## Handle Users

`GET` /users/{uid}

Gets the basic information of a user.

| Path Parameter | Description                           |
| -------------- | ------------------------------------- |
| uid            | _string_<br/> The unique ID of a user |

#### **Response Data**

See `UserRecord` in Firebase Admin SDK (Unnecessary properties will be removed, in the future).

`POST` /users/{uid}

Updates the basic information of a user. This endpoint is protected, only authorized requests are
handled. To authorize the request, calls `user.getIdToken()` to get the id token of the user, then
adds `Authorization: Bearer ${idToken}` to the header of the requests.

| Path Parameter | Description                           |
| -------------- | ------------------------------------- |
| uid            | _string_<br/> The unique ID of a user |

| Body Parameter         | Description                                |
| ---------------------- | ------------------------------------------ |
| displayName (optional) | _string_<br/> The display name of the user |
| photoURL (optional)    | _string_<br/> The user's photo URL         |

`DELETE` /users/{uid}

Deletes a user and _ALL_ information related to them. This endpoint is protected, only authorized
requests are handled. To authorize the request, calls `user.getIdToken()` to get the id token of the
user, then adds `Authorization: Bearer ${idToken}` to the header of the requests.

| Path Parameter | Description                           |
| -------------- | ------------------------------------- |
| uid            | _string_<br/> The unique ID of a user |

> Note: The following endpoints are not fully implemented; as the data models of sites, pages and
> comments may be changed for better performance/cleaner design.

`GET` /users/{uid}/sites

List all sites of a user. Only basic information will be sent back.

| Path Parameter | Description                           |
| -------------- | ------------------------------------- |
| uid            | _string_<br/> The unique ID of a user |

## Handle Sites

`GET` /sites/{siteId}

`POST` /sites/{siteId}

`DELTE` /sites/{siteId}

## Handle Pages

`GET` /pages/{pageId}

`POST` /pages/{pageId}

`DELTE` /pages/{pageId}

## Handle Comments (in pages)

`POST` /pages/{pageId}/comments/{commentId}

`DELTE` /pages/{pageId}/comments/{commentId}

**Reply by VietAnh1010** [reply]
Changes planned:
- `/pages/{pageId}` to `sites/{siteId}/pages/{pageId}` (same patern with comments). Maintains ONE nested collection only.
- `/pages/{pageId}/comments/{commentId}` to `comments/{commentId}`. Maintains THREE collections, no nested collection.

Mixed data models ... make me feel uncomfortable.

**Reply by joulev** [reply]
> POST /users/{uid}: Updates the basic information of a user.

Do not use `POST` to update information. Use `PUT` or `PATCH`. `POST` should only be used to *post* new information.

**Reply by joulev** [reply]
~~Also I think `GET /users/:uid` is not yet necessary at the moment, since after logging in with Firebase client, the user object is already available. Keep it though, who knows whether we will need it in the future.~~ nah nevermind

**Reply by joulev** [reply]
> All successful responses will contain a message and optional data, which is a JSON object,
or an JSON array in case multiple records are read.

I think I will infer the message from the status code (e.g. 200 means success, 500 means unknown error, 4xx means something's wrong on my end). The message field is therefore redundant.

**Reply by joulev** [reply]
@VietAnh1010 Further updates can be made [here](https://github.com/joulev/ezkomment/wiki/Internal-API-specs).

**Reply by VietAnh1010** [reply]
> ~Also I think `GET /users/:uid` is not yet necessary at the moment, since after logging in with Firebase client, the user object is already available. Keep it though, who knows whether we will need it in the future.~ nah nevermind

We will use this when the user extracts their information.

**Reply by joulev** [reply]
When Firebase signs someone in, `auth.currentUser` already contains all user information. So **for now** the route is not required. But do keep it.

**Reply by VietAnh1010** [reply]
Should the basic information of the members of an entity be sent back in response to `GET` requests?

For example, if a request is sent to `/api/sites/{siteId}`, an array contains the basic information of pages (does not contain information about individual comments in these pages) will be sent back, together with the basic properties the site (id, name, iconURL, etc.).

Current implementation does not do this.

**Reply by joulev** [reply]
Yes.
